Add Per solution funding

leerogers-bjss opened the pull request:
Add Per Solution funding controller, pages, validator and controller

Add OrderItemService for saving funding source information

Update RadioButtonTagHelper to accept a hint parameter as per the NHS Design

Update TableContentsTagHelper to add a Hint text section to the caption.

Fix capitalisation in TaskListTagHelper

Fix CommencementDateModel using system time for date conversion instead of treating everything as UTC

leerogers-bjss committed:
Add Per solution funding

Add Per Solution funding controller, pages, validator and controller

Add OrderItemService for saving funding source information

Update RadioButtonTagHelper to accept a hint parameter as per the NHS Design

Update TableContentsTagHelper to add a Hint text section to the caption.

Fix capitalisation in TaskListTagHelper

Fix CommencementDateModel using system time for date conversion instead of treating everything as UTC

bjss-colin commented on src/NHSD.GPIT.BuyingCatalogue.Services/Orders/OrderItemService.cs:
```diff
@@ -0,0 +1,88 @@
+
+            decimal totalCost = item.CalculateTotalCost();
+
+            switch (selectedFundingType)
```
Might be worth splitting this out to a method that returns a tuple `(double CentrallyAllocated, double LocallyAllocated)` for readability

leerogers-bjss replied:
nice call, worked well, done

bjss-colin commented on src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Validators/FundingSources/FundingSourceModelValidator.cs:
```diff
@@ -0,0 +1,37 @@
+            RuleFor(fs => fs.AmountOfCentralFunding)
+                .Cascade(CascadeMode.Stop)
+                .NotEmpty()
+                .When(fs => fs.SelectedFundingType == EntityFramework.Ordering.Models.OrderItemFundingType.MixedFunding)
```
Only a single `When` is required

> By default FluentValidation will apply the condition to all preceding validators in the same call to RuleFor
https://docs.fluentvalidation.net/en/latest/conditions.html#:~:text=By%20default%20FluentValidation%20will%20apply%20the%20condition%20to%20all%20preceding%20validators%20in%20the%20same%20call%20to%20RuleFor

leerogers-bjss replied:
oh nice, done

bjss-colin reviewed (changes requested):
Just a few syntax related comments.

Indentation is slightly ajar in the cshtml files too, might want to take a quick look. Razer support is quite bad in VS 2022.

leerogers-bjss commented on src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Ordering/Models/OrderItem.Custom.cs:
```diff
@@ -45,19 +45,19 @@ public override int GetHashCode()
+            };
         }
 
+        public bool ItemIsLocalFundingOnly() => CatalogueItem?.Solution?.FrameworkSolutions?.All(fs => fs.Framework.LocalFundingOnly) ?? false;
```
i'm currently rethinking this whole approach, since it doesn't work for additional or associated services due to their different structures.

leerogers-bjss replied:
fixed

leerogers-bjss commented on src/NHSD.GPIT.BuyingCatalogue.EntityFramework/Extensions/OrderItemFundingTypeExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace NHSD.GPIT.BuyingCatalogue.EntityFramework.Extensions
+{
+    public static class OrderItemFundingTypeExtensions
```
going to do this as a seperate tech debt PR

leerogers-bjss committed:
Update Task list logic

Update the taskListService to not require the full order object to work out its status but to get the information itself.

Update OrderItemService to get the centrally/locally allocated values correctly.

bjss-colin commented on src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/FundingSource/FundingSources.cshtml:
```diff
@@ -0,0 +1,116 @@
+		{
+			<nhs-details label-text="What is central, local and mixed funding?">
+				<h2 class="nhsuk-heading-m">Central funding</h2>
+				<p>Central funding means you’re using your centrally held allocation to pay for the order. Any solutions or services you order will be paid for by either NHS England or NHS Digital. The supplier will invoice the relevant organisation directly.</p>
```
There's a unicode character which is causing problems with syntax highlighting

> Central funding means you’re using your centrally held allocation to pay for the order

The symbol in the word `you're` is a [right single quotation mark](https://en.wikipedia.org/wiki/Right_single_quotation_mark) instead of being an apostrophe

leerogers-bjss replied:
done

bjss-colin commented on tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/TaskList/TaskListServiceTests.cs:
```diff
@@ -10,14 +15,25 @@ namespace NHSD.GPIT.BuyingCatalogue.Services.UnitTests.TaskList
 {
     public static class TaskListServiceTests
     {
+        /* TODO - Reimplement Task list Unit Tests
```
Given this change has been done as part of this PR, I think these should be implemented within the scope of this PR also

leerogers-bjss replied:
this will be done along side the Unit tests i need to do for this work, i just put it in as a todo to remind myself

bjss-colin replied:
Any reason the unit tests couldn't be included with the work? Ideally every PR should have the complimentary unit tests. Putting something like the End to End tests as a separate PR is reasonable since we'd have a level of coverage already with unit tests.

I appreciate this isn't going directly to the develop branch and we can add them later, which is why I'm inclined to approve this rather than put up much of a battle, but I don't think this is an approach we should be getting comfortable with.

leerogers-bjss committed:
replace right single quotation mark with paostophes
